### PR TITLE
🌱 Compute desired Cluster and its referenced objects for a managed topology

### LINKED ISSUE #4998: Implement generate Cluster and its referenced objects for a managed topology
**Detailed Description**

This is part of the activities for the implementation of the [ClusterClass proposal](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/202105256-cluster-class-and-managed-topologies.md). 

As part of the ManageTopologyController reconcile loop, we want to generate the desired state of a managed topology.

This requires to computing following objects starting from the ClusterClass, referenced templates, and eventually the current state of the cluster:

- ClusterInfrastructure object
- ControlPlaneMachineInfrastructureTemplate object, if required
- ControlPlane object
- Cluster object (only spec.InfrastructureRef and spec.ControlPlaneRef are computed)
- For each MachineDeployment in Cluster.Topology.Workers.Deployments
   - MachineBootstrapTemplate object
   - MachineInfrastructureTemplate object
   - MachineDeployment object

Given the scope of this task, this issue focus on the first set of objects, cluster and its dependencies, while machine deployments and related templates will be addressed in a separated issue.

Also, please note that implementing support for version upgrade component by component will be addressed in a follow up step.

** Anything else you would like to add: ** 

This issue assumes https://github.com/kubernetes-sigs/cluster-api/issues/4933 and https://github.com/kubernetes-sigs/cluster-api/pull/4986 to be merged first

The rules for creating objects are documented in the proposal, however some things are worth to notice:
- Creating a ClusterInfrastructure object from a template (template .--> object) is really similar what we are already doing when creating a MachineInfrastructure object from a template; same considerations apply to creating the ControlPlane object
- Creating the ControlPlaneMachineInfrastructureTemplate is slightly different (template .--> template), but we want to keep a consistent set of labels and annotations (e.g ClonedFrom)
- Crating ControlPlane should take into account providers with/without replicas support and providers with/without requirement for a MachineInfrastructureTemplate

/kind feature

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:

This is part of the activities for the implementation of the [ClusterClass proposal](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/202105256-cluster-class-and-managed-topologies.md). 

This PR computes the desired state for the Cluster and its referenced objects.

WIP: this PR is based on https://github.com/kubernetes-sigs/cluster-api/pull/4986; please consider only the last commit

**Which issue(s) this PR fixes**:
Fixes #4998

### CODE REVIEW

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:162]:
```suggestion
		return nil, errors.Wrap(err, "failed to generate the ControlPlane object")
```
In case we want to align the errors to this. (would apply to a few more errors below too)

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:141]:
```suggestion
		templateClonedFromRef: class.clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref,
```
after #5000, right? (mostly to confirm my understanding)

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:144]: I guess after the new version of the `getNestedRef` func is used we will handle the error. I guess when the ref is not already set we would have to ignore the error here (which we get when the ref does not exist)

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:209]:
```suggestion
	cluster.Labels[clusterv1.ClusterTopologyLabelName] = cluster.Name
```
according to the proposal

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:219]:
```suggestion
type templateToObjectInput struct {
```
nit (if you want, I think that was our pattern in other cases)

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [reply controllers/clustertopology_controller_compute.go:209]: Same below (2x)

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:255]: The msg needs parameterization as it's also used in other contexts

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:272]:
```suggestion
// NOTE: We are creating a copy of the InfrastructureCluster template for each cluster so
```

Comment (sbueringer) [reply controllers/clustertopology_controller_compute.go:272]: Although I'm not sure if this comment isn't to specific given that this func is used in multiple contexts

**Review by sbueringer — COMMENTED** [review]
Great work!!!
Only nits. I would give it another review after the two underlying PRs have been merged.

File: controllers/clustertopology_controller_compute.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:152]: yes!

File: controllers/clustertopology_controller_compute.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go]: outdated

File: controllers/clustertopology_controller_compute.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:221]: I have changed the proposal, there is already a cluster name label on all the objects, no need to repeat it...

File: controllers/clustertopology_controller_compute.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:231]: This input struct is used by two func, also it is internal

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [reply controllers/clustertopology_controller_compute.go:221]: Agree :)

**Comment by sbueringer** [comment]
lgtm apart from the one TODO after #4991 is merged

**Comment by fabriziopandini** [comment]
@sbueringer TODO addressed
@vincepri this one should be ready to go...

File: controllers/clustertopology_controller_compute_test.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute_test.go:627]:
```suggestion
func assertNestedFieldDoesNotExist(g *WithT, obj *unstructured.Unstructured, fields ...string) {
```

Comment (sbueringer) [reply controllers/clustertopology_controller_compute_test.go:627]: Or alternatively if you like `assertNestedFieldUnset`

**Review by sbueringer — COMMENTED** [review]
@fabriziopandini one nit

File: controllers/clustertopology_controller_compute_test.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute_test.go]: Done

**Comment by fabriziopandini** [comment]
rebased...

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**
